Cds 89396 (#6247)
harness/developer-hub: kb/continuous-delivery/articles/servicenow-api-permission.md

The `bearer token` mentioned above is the token generated by the credentials provided in ServiceNow.

Required permission: `itil`.

### Update multiple ServiceNow tickets without templates

### Update a ServiceNow ticket from templates

Harness uses custom script to call ServiceNow APIs to update a ticket from templates.
Required permission: `x_harne_harness_ap.integration_user_role`.

## ServiceNow Approval step

Harness uses the following API to get the fields of the ServiceNow ticket for which the approval or rejection criteria is specified in the Approval step.

```
curl --location --request GET '{instance_url}/api/now/table/incident?sysparm_display_value=all&sysparm_query=number=INC0011331' \
--header 'Authorization: Basic {bearer token}' \
```

The `instance_url` mentioned above is the ServiceNow URL provided in the ServiceNow connector.
The `bearer token` mentioned above is the token generated by the credentials provided in ServiceNow.

Required permission: `itil`.

## General debugging FAQs

1. **Why is apply from template selection blocked when using ServiceNow create/update step?**

>    The connector provided is probably missing the `x_harne_harness_ap.integration_user_role` role permission. Another possible reason could be that the delegate might be down.

Author: ```suggestion
   The connector provided is probably missing the `x_harne_harness_ap.integration_user_role` role permission. Another possible reason could be that the delegate might be down. Also, check if Integration for Harness Software Delivery Platform app is installed in your ServiceNow instance from the ServiceNow store before using templates to update ServiceNow tickets. For more information, go to [Integration for Harness Software Delivery Platform](https://store.servicenow.com/sn_appstore_store.do#!/store/application/de154a1e1b75851044cbdb58b04bcb11/1.0.2?referer=%2Fstore%2Fsearch%3Flistingtype%3Dallintegrations%25253Bancillary_app%25253Bcertified_apps%25253Bcontent%25253Bindustry_solution%25253Boem%25253Butility%25253Btemplate%26q%3Dharness&sl=sh).
```

---
description: KB - Debug ServiceNow create, update, and approval steps
title: ServiceNow create, update, and approval API permissions
---

> This KB discusses the permissions required for ServiceNow create, update, and approval steps.

Reviewer: I think we should link to relevant reference docs here:. RBAC, ServiceNow step reference sheets, etc. 

what do you think?

## Create ServiceNow tickets

### Create a ServiceNow ticket without templates

Harness uses the following API to create a ServiceNow ticket without templates:  

```

> curl --location --request POST '{instance_url}/api/now/table/incident?sysparm_display_value=all' \

Author: @naman-goenka @VivekHarness could you please help with the field details?

> curl --location --request POST '{instance_url}/api/now/table/incident?sysparm_display_value=all' \

Author: ```suggestion
curl --location --request POST '{instance_url}/api/now/table/<ticket_type>'\
```

--header 'Authorization: Basic {bearer token}' \
--header 'Content-Type: application/json' \
--data-raw '{
    "description" : "sample incident"
}'
```

The `instance_url` mentioned above is the ServiceNow URL provided in the ServiceNow connector.
The `bearer token` mentioned above is the token generated by the credentials provided in ServiceNow.

> Required permission: `itil`.

Reviewer: I'm assuming these permissions are for [RBAC](https://developer.harness.io/docs/platform/role-based-access-control/rbac-in-harness/)

We should link to the documentation for it somewhere in the KB.

Reviewer 2: This is not related to harness rbac, but rbac of integration user in serviceNow.

> 

Reviewer 2: @vivekkarnatiharness is itil enough for fetching the metadata as well in create step? Else we need to specify the same here.

Reviewer 3: checking

### Create a ServiceNow ticket from a form templates

Harness uses custom script APIs to create a ticket using form templates.

> Required permission: `itil`, `x_harne_harness_ap.integration_user_role`.

Reviewer 2: @vivekkarnatiharness @roshnisarangadharan can we add a note here/cross-link that this flow requires harness app to be installed in servicenow, so that it helps in debugging. The same in present in step docs as well.
https://developer.harness.io/docs/continuous-delivery/x-platform-cd-features/cd-steps/ticketing-systems/create-service-now-tickets-in-cd-stages/#important-notes-for-using-templates

### Create a ServiceNow ticket from a standard templates

> This is supported only for change request task types.

Reviewer 2: change request `ticket` types.

Reviewer 2: @vivekkarnatiharness @roshnisarangadharan we need to add create API used here and the permissions required. 
Note: change_manager is required to get readOnly fields in standard change template. 
Note: is itil sufficient for getting templates as well?

> This is supported only for change request task types.

Author: ```suggestion
This is supported only for change request ticket types.
```

> Harness uses custom script APIs to create a ticket using form template.

Reviewer 2: This seems redundant here

Reviewer 2: this needs to be fixed

Reviewer 2: The perms are also incorrect here

Required permission: `itil`, `x_harne_harness_ap.integration_user_role`.

## Update ServiceNow tickets 

### Update a ServiceNow ticket without templates

Harness uses the following API to update a ServiceNow ticket without templates:  

```

> curl --location --request PATCH '{instance_url}/api/now/table/incident/57bea4ef1b41e150c312755e034bcbd1?sysparm_display_value=all' \

Reviewer 2: we can templatize table_name to the ticket type field as well

--header 'Authorization: Basic {bearer token}' \
--header 'Content-Type: application/json' \
--data-raw '{
    "description" : "updated"
}'
```

The `instance_url` mentioned above is the ServiceNow URL provided in the ServiceNow connector.
The `bearer token` mentioned above is the token generated by the credentials provided in ServiceNow.

Required permission: `itil`.

### Update multiple ServiceNow tickets without templates

> This functionality is behind the feature flag, `CDS_NG_UPDATE_MULTIPLE_SNOW_CHANGE_REQUEST`. Contact [Harness Support](mailto:support@harness.io) to enable this functionality.

Reviewer: Do we have docs on how to use this feature we could link or mention here?

Reviewer 2: This flag is GAed

> This functionality is behind the feature flag, `CDS_NG_UPDATE_MULTIPLE_SNOW_CHANGE_REQUEST`. Contact [Harness Support](mailto:support@harness.io) to enable this functionality.

Author: ```suggestion

```

### Update a ServiceNow ticket from templates

Harness uses custom script to call ServiceNow APIs to update a ticket from templates.

> Required permission: `x_harne_harness_ap.integration_user_role`.

Reviewer 2: is `itil` not required here, for listing templates as well?

Author: @vivekkarnatiharness please confirm

Reviewer 3: itil role is required here, will update the pr with that change

Reviewer 2: this doesn't seems to be added

Reviewer 2: Can we mention the same here (requirement to have app installed)

## ServiceNow Approval step

Harness uses the following API to get the fields of the ServiceNow ticket for which the approval or rejection criteria is specified in the Approval step.

```

> curl --location --request GET '{instance_url}/api/now/table/incident?sysparm_display_value=all&sysparm_query=number=INC0011331' \

Reviewer 2: we can templatize table_name to the ticket type field as well

--header 'Authorization: Basic {bearer token}' \
```

The `instance_url` mentioned above is the ServiceNow URL provided in the ServiceNow connector.
The `bearer token` mentioned above is the token generated by the credentials provided in ServiceNow.

Required permission: `itil`.

## General debugging FAQs

1. **Why is apply from template selection blocked when using ServiceNow create/update step?**

> 

Reviewer 2: app might not have been installed as well

Author: Please check if Integration for Harness Software Delivery Platform app is installed in your ServiceNow instance from the ServiceNow store before using templates to update ServiceNow tickets.
[link](https://store.servicenow.com/sn_appstore_store.do#!/store/application/de154a1e1b75851044cbdb58b04bcb11/1.0.2?referer=%2Fstore%2Fsearch%3Flistingtype%3Dallintegrations%25253Bancillary_app%25253Bcertified_apps%25253Bcontent%25253Bindustry_solution%25253Boem%25253Butility%25253Btemplate%26q%3Dharness&sl=sh)

   The connector provided is probably missing the `x_harne_harness_ap.integration_user_role` role permission. Another possible reason could be that the delegate might be down.

> 2. **Why does updating a ServiceNow ticket from template results in an error, `Error in fetching issue 125c716edb61a910e9748c3b13961918 .Issue does not exist`.**

Reviewer: ```suggestion
2. **Why does updating a ServiceNow ticket from a template result in the error: `Error in fetching issue 125c716edb61a910e9748c3b13961918 .Issue does not exist`.**
```

   The selected ticket type might be different from the actual ticket.

> 3. **How to debug ACL error?**

Reviewer: ```suggestion
3. **How do you debug an ACL error?**
```

   Assume that the integration user has all above mentioned roles for specific use cases.

   1. Identify the user or integration that Harness used as a connector. This could be a specific user, system account, or an integration like an API. 
   2. ServiceNow uses Access Control Lists (ACLs) to control what data a user can access and what they can do with that data. Check the ACLs for the table to which you are trying to insert/update. For example, Incident.
   3. Ensure that the user or integration has the necessary roles to perform the action. This might involve checking their roles, groups, or specific ACL rules. 
   4. If the user or integration does not have the necessary permissions, you will need to update the ACLs. This could involve adding a new ACL rule, or modifying an existing one.
   5. Once the ACLs are updated, create the ServiceNow ticket again. If the insert is successful, the issue has been resolved.

>    6. ServiceNow has data restrictions on their UI. You can use the APIs listed above for debugging.

Reviewer 2: There can be various UI policies configured on ServiceNow, that may not allow to perform the operation directly on UI, even is that is possible via Harness. You can directly use the APIs listed above for debugging permission issues.

:::

### Create a ServiceNow ticket from a standard templates

This is supported only for change request ticket types.

Required permission: `itil`, `x_harne_harness_ap.integration_user_role`.

## Update ServiceNow tickets 

### Update a ServiceNow ticket without templates

Harness uses the following API to update a ServiceNow ticket without templates:  

```
curl --location --request PATCH '{instance_url}/api/now/table/<ticket_type>' \
--header 'Authorization: Basic {bearer token}' \
--header 'Content-Type: application/json' \
--data-raw '{
    "description" : "updated"
}'
```

The `instance_url` mentioned above is the ServiceNow URL provided in the ServiceNow connector.
The `bearer token` mentioned above is the token generated by the credentials provided in ServiceNow.

Required permission: `itil`.

### Update multiple ServiceNow tickets without templates

### Update a ServiceNow ticket from templates

> Harness uses custom script to call ServiceNow APIs to update a ticket from templates.

Author: @vivekkarnatiharness I need the APIs. Could you please add here?

```
curl --location --request POST '{instance_url}/api/now/table/<ticket_type>'\
--header 'Authorization: Basic {bearer token}' \
--header 'Content-Type: application/json' \
--data-raw '{
    "description" : "sample incident"
}'
```

The `instance_url` mentioned above is the ServiceNow URL provided in the ServiceNow connector.
The `bearer token` mentioned above is the token generated by the credentials provided in ServiceNow.

Required permission: `itil`.

### Create a ServiceNow ticket from a form templates

Harness uses custom script APIs to create a ticket using form templates.
Required permission: `itil`, `x_harne_harness_ap.integration_user_role`.

:::note 
This flow requires the Harness app to be installed in ServiceNow, so that it helps in debugging. Review [Important notes for using templates](https://developer.harness.io/docs/continuous-delivery/x-platform-cd-features/cd-steps/ticketing-systems/create-service-now-tickets-in-cd-stages/#important-notes-for-using-templates) for more details.
:::

### Create a ServiceNow ticket from a standard templates

This is supported only for change request ticket types.

> Harness uses custom script APIs to create a ticket using form template.

Author: ```suggestion

```

Required permission: `itil`, `x_harne_harness_ap.integration_user_role`.

## Update ServiceNow tickets 

### Update a ServiceNow ticket without templates

Harness uses the following API to update a ServiceNow ticket without templates:  

```
curl --location --request PATCH '{instance_url}/api/now/table/<ticket_type>' \
--header 'Authorization: Basic {bearer token}' \
--header 'Content-Type: application/json' \
--data-raw '{
    "description" : "updated"
}'
```

The `instance_url` mentioned above is the ServiceNow URL provided in the ServiceNow connector.
The `bearer token` mentioned above is the token generated by the credentials provided in ServiceNow.

Required permission: `itil`.

### Update multiple ServiceNow tickets without templates

### Update a ServiceNow ticket from templates

Harness uses custom script to call ServiceNow APIs to update a ticket from templates.

> Required permission: `x_harne_harness_ap.integration_user_role`.

Author: ```suggestion
Required permission: `itil_admin`, `x_harne_harness_ap.integration_user_role`.
```

Required permission: `itil`.

## General debugging FAQs

1. **Why is apply from template selection blocked when using ServiceNow create/update step?**

   The connector provided is probably missing the `x_harne_harness_ap.integration_user_role` role permission. Another possible reason could be that the delegate might be down. Also, check if Integration for [Harness Software Delivery Platform app](https://store.servicenow.com/sn_appstore_store.do#!/store/application/de154a1e1b75851044cbdb58b04bcb11/1.0.2?referer=%2Fstore%2Fsearch%3Flistingtype%3Dallintegrations%25253Bancillary_app%25253Bcertified_apps%25253Bcontent%25253Bindustry_solution%25253Boem%25253Butility%25253Btemplate%26q%3Dharness&sl=sh) is installed in your ServiceNow instance from the ServiceNow store before using templates to update ServiceNow tickets. 

2. **Why does updating a ServiceNow ticket from a template result in the error: `Error in fetching issue 125c716edb61a910e9748c3b13961918 .Issue does not exist`.**

   The selected ticket type might be different from the actual ticket.

3. **How do you debug an ACL error?**

>    Assume that the integration user has all above mentioned roles for specific use cases.

Reviewer 2: assumin*

>    Assume that the integration user has all above mentioned roles for specific use cases.

Author: ```suggestion
   Assuming that the integration user has all above mentioned roles for specific use cases.
```

> ---

Reviewer 2: Please ensure the we take into consideration permissions for `ticket` API and `getMetadata` API as they are used by FE in each of these steps. We any permission apart from `itil` is required we can mention it in separate section as well.

description: KB - Debug ServiceNow create, update, and approval steps
title: ServiceNow create, update, and approval API permissions
---

This KB discusses the permissions required for ServiceNow create, update, and approval steps.

## Create ServiceNow tickets

### Create a ServiceNow ticket without templates

Harness uses the following API to create a ServiceNow ticket without templates:  

```
curl --location --request POST '{instance_url}/api/now/table/<ticket_type>'\
--header 'Authorization: Basic {bearer token}' \
--header 'Content-Type: application/json' \
--data-raw '{
    "description" : "sample incident"
}'
```

The `instance_url` mentioned above is the ServiceNow URL provided in the ServiceNow connector.
The `bearer token` mentioned above is the token generated by the credentials provided in ServiceNow.

Required permission: `itil`.

### Create a ServiceNow ticket from a form templates

Harness uses custom script APIs to create a ticket using form templates.

> Required permission: `itil`.

Author: ```suggestion
Required permission: `itil`, `x_harne_harness_ap.integration_user_role`.
```

:::note 

> This flow requires the Harness app to be installed in ServiceNow, so that it helps in debugging. Review [Important notes for using templates](https://developer.harness.io/docs/continuous-delivery/x-platform-cd-features/cd-steps/ticketing-systems/create-service-now-tickets-in-cd-stages/#important-notes-for-using-templates) for more details.

Reviewer 2: pls remove `so that it helps in debugging.`

> This flow requires the Harness app to be installed in ServiceNow, so that it helps in debugging. Review [Important notes for using templates](https://developer.harness.io/docs/continuous-delivery/x-platform-cd-features/cd-steps/ticketing-systems/create-service-now-tickets-in-cd-stages/#important-notes-for-using-templates) for more details.

Author: ```suggestion
This flow requires the Harness app to be installed in ServiceNow. Review [Important notes for using templates](https://developer.harness.io/docs/continuous-delivery/x-platform-cd-features/cd-steps/ticketing-systems/create-service-now-tickets-in-cd-stages/#important-notes-for-using-templates) for more details.
```

:::

### Create a ServiceNow ticket from a standard templates

> This is supported only for change request ticket types.

Author: ```suggestion
This is supported only for change request ticket types. 

While fetching fields, if the `change_manager` permission is not available, then Harness fetches only the read-only fields mentioned `description`, `backout_plan`, `test_plan`, and `implementation_plan`.
```

> While fetching fields, if the `change_manager` permission is not available, then Harness fetches only the read-only fields mentioned `description`, `backout_plan`, `test_plan`, and `implementation_plan`.

Reviewer 2: Harness consideres following fields as readOnly ... instead of adhering to ServiceNow properties

> While fetching fields, if the `change_manager` permission is not available, then Harness fetches only the read-only fields mentioned `description`, `backout_plan`, `test_plan`, and `implementation_plan`.

Author: ```suggestion
While fetching fields, if the `change_manager` permission is not available, then Harness considers the `description`, `backout_plan`, `test_plan`, and `implementation_plan` fields as read-only.
```

> 

Reviewer 2: We should add the API here

Reviewer 2: ```
curl --location --request POST '{instance_url}/api/sn_chg_rest/change/standard/{sys_id}?backout_plan=ABORT&test_plan=TEST'\
--header 'Authorization: Basic {bearer token}' \
--header 'Content-Type: application/json' \
```

> Required permission: `itil`.

Reviewer 2: add `change_manager` here

> Required permission: `itil`, `x_harne_harness_ap.integration_user_role`.

Author: ```suggestion
Required permission: `itil`.
```

> 

Reviewer 2: Pls remove `x_harne_harness_ap.integration_user_role`

## Update ServiceNow tickets 

### Update a ServiceNow ticket without templates

Harness uses the following API to update a ServiceNow ticket without templates:  

```
curl --location --request PATCH '{instance_url}/api/now/table/<ticket_type>' \
--header 'Authorization: Basic {bearer token}' \
--header 'Content-Type: application/json' \
--data-raw '{
    "description" : "updated"
}'
```

The `instance_url` mentioned above is the ServiceNow URL provided in the ServiceNow connector.
The `bearer token` mentioned above is the token generated by the credentials provided in ServiceNow.

Required permission: `itil`.

### Update a ServiceNow ticket from templates

Harness uses custom script to call ServiceNow APIs to update a ticket from templates.
Required permission: `itil_admin`, `x_harne_harness_ap.integration_user_role`.

> Required role: `change_manager`.

Reviewer 2: delete 70-78

   3. Ensure that the user or integration has the necessary roles to perform the action. This might involve checking their roles, groups, or specific ACL rules. 
   4. If the user or integration does not have the necessary permissions, you will need to update the ACLs. This could involve adding a new ACL rule, or modifying an existing one.
   5. Once the ACLs are updated, create the ServiceNow ticket again. If the insert is successful, the issue has been resolved.
   6. ServiceNow has data restrictions on their UI. There can be various UI policies configured on ServiceNow that may not allow you to perform the operation directly on UI even it is possible via Harness. You can directly use the APIs listed above for debugging permission issues.
   7. Check if the table, for example Change Request has any scoped applications enabled. For details, go to [ServiceNow documentation](https://docs.servicenow.com/bundle/vancouver-application-development/page/build/applications/reference/r_TableApplicationAccessFields.html).
      ![image](../static/debug-approval-step.png)

>    8. Check if there are any data policies written on the user's Dev instance on the table. For details, go to [ServiceNow documentation](https://docs.servicenow.com/bundle/vancouver-platform-administration/page/administer/field-administration/concept/c_DataPolicy.html).

Reviewer 2: 9. When using custom tables, i.e., not Incident, Problem, Change Request, and Change Task in create/update using form templates, please ensure to grant cross-scope privileges for the required table to harness app in servicenow. More details: 

https://developer.harness.io/docs/continuous-delivery/x-platform-cd-features/cd-steps/ticketing-systems/create-service-now-tickets-in-cd-stages#custom-table-support:~:text=the%20ServiceNow%20store.-,%3A%3A%3A%5Cnotes,-Cross%2Dscope%20privileges

---
description: KB - Debug ServiceNow create, update, and approval steps
title: ServiceNow create, update, and approval API permissions
---

This KB discusses the permissions required for ServiceNow create, update, and approval steps.

## Create ServiceNow tickets

### Create a ServiceNow ticket without templates

Harness uses the following API to create a ServiceNow ticket without templates:  

```
curl --location --request POST '{instance_url}/api/now/table/<ticket_type>'\
--header 'Authorization: Basic {bearer token}' \
--header 'Content-Type: application/json' \
--data-raw '{
    "description" : "sample incident"
}'
```

The `instance_url` mentioned above is the ServiceNow URL provided in the ServiceNow connector.
The `bearer token` mentioned above is the token generated by the credentials provided in ServiceNow.

Required permission: `itil`.

### Create a ServiceNow ticket from a form templates

Harness uses custom script APIs to create a ticket using form templates.

> Required permission: `itil`, `x_harne_harness_ap.integration_user_role`.

Author: ```suggestion
Required permission: `itil`.
```

:::note 
This flow requires the Harness app to be installed in ServiceNow, so that it helps in debugging. Review [Important notes for using templates](https://developer.harness.io/docs/continuous-delivery/x-platform-cd-features/cd-steps/ticketing-systems/create-service-now-tickets-in-cd-stages/#important-notes-for-using-templates) for more details.
:::

### Create a ServiceNow ticket from a standard templates

This is supported only for change request ticket types.

> Required permission: `itil`, `x_harne_harness_ap.integration_user_role`.

Author: While fetching fields, if the permission (i.e. change_manager) is not available, Then we will default to the list of read-only fields mentioned "description", "backout_plan", "test_plan", "implementation_plan"

Harness uses the following API to update a ServiceNow ticket without templates:  

```
curl --location --request PATCH '{instance_url}/api/now/table/<ticket_type>' \
--header 'Authorization: Basic {bearer token}' \
--header 'Content-Type: application/json' \
--data-raw '{
    "description" : "updated"
}'
```

The `instance_url` mentioned above is the ServiceNow URL provided in the ServiceNow connector.
The `bearer token` mentioned above is the token generated by the credentials provided in ServiceNow.

Required permission: `itil`.

### Update multiple ServiceNow tickets without templates

### Update a ServiceNow ticket from templates

Harness uses custom script to call ServiceNow APIs to update a ticket from templates.
Required permission: `itil_admin`, `x_harne_harness_ap.integration_user_role`.
Required role: `change_manager`.

Harness uses the following API to update a ServiceNow ticket from templates:

```
@Headers("Content-Type: application/json")
@POST("api/sn_chg_rest/change/standard/{sys_id}?")
Call<JsonNode> createTicketUsingStandardTemplate(@Header("Authorization") String authorization,
    @Path("sys_id") String sys_id, @QueryMap Map<String, String> queryParams);

> ```

Author: @naman-goenka @vivekkarnatiharness please check if this is accurate.

> ```
> @Headers("Content-Type: application/json")
> @POST("api/sn_chg_rest/change/standard/{sys_id}?")
> Call<JsonNode> createTicketUsingStandardTemplate(@Header("Authorization") String authorization,
>     @Path("sys_id") String sys_id, @QueryMap Map<String, String> queryParams);
> ```

Author: ```suggestion
```
curl --location --request POST '{instance_url}/api/sn_chg_rest/change/standard/{sys_id}?backout_plan=ABORT&test_plan=TEST'\
--header 'Authorization: Basic {bearer token}' \
--header 'Content-Type: application/json' \
```
```

:::

### Create a ServiceNow ticket from a standard templates

This is supported only for change request ticket types. 

While fetching fields, if the `change_manager` permission is not available, then Harness fetches only the read-only fields mentioned `description`, `backout_plan`, `test_plan`, and `implementation_plan`.

Required permission: `itil`, `x_harne_harness_ap.integration_user_role`.

## Update ServiceNow tickets 

### Update a ServiceNow ticket without templates

Harness uses the following API to update a ServiceNow ticket without templates:  

```
curl --location --request PATCH '{instance_url}/api/now/table/<ticket_type>' \
--header 'Authorization: Basic {bearer token}' \
--header 'Content-Type: application/json' \
--data-raw '{
    "description" : "updated"
}'
```

The `instance_url` mentioned above is the ServiceNow URL provided in the ServiceNow connector.
The `bearer token` mentioned above is the token generated by the credentials provided in ServiceNow.

Required permission: `itil`.

> ### Update multiple ServiceNow tickets without templates

Author: @vivekkarnatiharness need info here

Harness uses the following API to create a ServiceNow ticket without templates:  

```
curl --location --request POST '{instance_url}/api/now/table/<ticket_type>'\
--header 'Authorization: Basic {bearer token}' \
--header 'Content-Type: application/json' \
--data-raw '{
    "description" : "sample incident"
}'
```

The `instance_url` mentioned above is the ServiceNow URL provided in the ServiceNow connector.
The `bearer token` mentioned above is the token generated by the credentials provided in ServiceNow.

Required permission: `itil`.

### Create a ServiceNow ticket from a form templates

Harness uses custom script APIs to create a ticket using form templates.
Required permission: `itil`, `x_harne_harness_ap.integration_user_role`.

:::note 
This flow requires the Harness app to be installed in ServiceNow. Review [Important notes for using templates](https://developer.harness.io/docs/continuous-delivery/x-platform-cd-features/cd-steps/ticketing-systems/create-service-now-tickets-in-cd-stages/#important-notes-for-using-templates) for more details.
:::

### Create a ServiceNow ticket from a standard templates

This is supported only for change request ticket types. 

> While fetching fields, if the `change_manager` permission is not available, then Harness considers the `description`, `backout_plan`, `test_plan`, and `implementation_plan` fields as read-only. 

Author: ```suggestion
While fetching fields, if the `change_manager` permission is not available, then Harness considers the `description`, `backout_plan`, `test_plan`, and `implementation_plan` fields as read-only instead of adhering to ServiceNow properties. 
```

Required permission: `itil`.

> Required role: `change_manager`.

Author: ```suggestion

Required role: `change_manager`.
```
